File: conformance/bundles/02-context-isolation/counter.steps.ts
import { steps } from '@varar/varar'

const { stimulus, sensor } = steps<{ count: number }>(() => ({ count: 0 }))

stimulus('I increment', (state) => ({ count: state.count + 1 }))

// One slot ({int}): return the observed count and the core compares it against
// the number in the document.
sensor('The count is {int}', (state) => state.count)

File: conformance/bundles/02-context-isolation/counter.steps.cs
// C# sibling of counter.steps.ts / .rs (bundle 02-context-isolation).
using Varar;
using Varar.Core;

namespace Varar.Corpus.B02;

public static class CounterSteps
{
    public static void Register(Steps s)
    {
        s.Stimulus("I increment", state => Value.Map([new("count", Value.Of(CountOf(state) + 1))]));

        // One slot ({int}): return the observed count and let the core compare it against the
        // number in the document.
        s.Sensor("The count is {int}", (state, n) => Value.Of(CountOf(state)));
    }

    public static Value State() => Value.Map([new("count", Value.Of(0))]);

    private static long CountOf(Value state) =>
        state is VMap m && m.Entries.TryGetValue("count", out var v) && v is VInt i ? i.Int : 0;
}
